code/ship/ship.cpp

```diff
@@ -8856,10 +8856,76 @@ static void lethality_decay(ai_info *aip)
 #endif
 }
 
-void ship_process_pre(object *objp, float frametime)
+// moved out of ship_process_post() so it can be called from either -post() or -pre() depending on Ai_before_physics
+void ship_evaluate_ai(object* obj, float frametime) {
+
+	int num = obj->instance;
+	Assert(num >= 0 && num < MAX_SHIPS);
+	Assert(Ships[num].objnum == OBJ_INDEX(obj));
+
+	ship* shipp = &Ships[num];
+
+	//rotate player subobjects since its processed by the ai functions
+	// AL 2-19-98: Fire turret for player if it exists
+	//WMC - changed this to call process_subobjects
+	if ((obj->flags[Object::Object_Flags::Player_ship]) && !Player_use_ai)
+	{
+		ai_info* aip = &Ai_info[Ships[obj->instance].ai_index];
+		if (aip->ai_flags[AI::AI_Flags::Being_repaired, AI::AI_Flags::Awaiting_repair])
+		{
+			if (aip->support_ship_objnum >= 0)
+			{
+				if (vm_vec_dist_quick(&obj->pos, &Objects[aip->support_ship_objnum].pos) < (obj->radius + Objects[aip->support_ship_objnum].radius) * 1.25f)
+					return;
+			}
+		}
+		if (!shipp->flags[Ship::Ship_Flags::Rotators_locked])
+			process_subobjects(OBJ_INDEX(obj));
+	}
+
+	// update ship lethality
+	if (Ships[num].ai_index >= 0 ){
+		if (!physics_paused && !ai_paused) {
+			lethality_decay(&Ai_info[Ships[num].ai_index]);
+		}
+	}
+
+	// if the ship is an observer ship don't need to do AI
+	if ( obj->type == OBJ_OBSERVER)  {
+		return;
+	}
+
+	// Goober5000 - player may want to use AI
+	if ( (Ships[num].ai_index >= 0) && (!(obj->flags[Object::Object_Flags::Player_ship]) || Player_use_ai) ){
+		if (!physics_paused && !ai_paused){
+			ai_process( obj, Ships[num].ai_index, frametime );
+		}
+	}
+}
+
+void ship_process_pre(object *obj, float frametime)
 {
-	if ( (objp == NULL) || !frametime )
+	// If Ai_before_physics is false everything following is evaluated in ship_process_post()
+	// Also only multi masters do ai
+	if ( (obj == nullptr) || !frametime || MULTIPLAYER_CLIENT || !Ai_before_physics)
+		return;
+
+	int num = obj->instance;
+	Assert(num >= 0 && num < MAX_SHIPS);
+	Assert(Ships[num].objnum == OBJ_INDEX(obj));
+	ship* shipp = &Ships[num];
+
+	if (obj->type != OBJ_SHIP) {
+		nprintf(("AI", "Ignoring non-ship object in ship_process_pre()\n"));
 		return;
+	}
+
+	if ((!(shipp->is_arriving()) || (Ai_info[shipp->ai_index].mode == AIM_BAY_EMERGE)
+		|| ((Warp_params[shipp->warpin_params_index].warp_type == WT_IN_PLACE_ANIM) && (shipp->flags[Ship_Flags::Arriving_stage_2])))
+		&& !(shipp->flags[Ship_Flags::Depart_warp]))
```

Comment:
These conditions don't appear to be part of the original code

Reply:
They are, you can find them at line 9071 on the current master (9112 on this PR). They're just kind of spaced somewhat away from the AI stuff because it uses the same conditions for some other non-AI stuff.

Reply:
Ah, ok, thanks.